### ISSUE TITLE
Unable to import/restore backup with interleaved indexes

### ISSUE THREAD

**Issue by roncrdb** [cockroachdb/cockroach#38310]
**Describe the problem**
Unimarket has run into a unique issue.  They have a full backup that was created, they have a few interleaved tables and indexes. However when they try to restore using a command such as: 

```
root@:26257/defaultdb> RESTORE database <DB> from 'nodelocal:///anz-cluster-demo-staging-catalogue-partitioned.sql';
pq: importing 12 ranges: could not rewrite span start key: /Table/459/1/"NZ"/PrefixEnd: did not find terminator 0x0 in buffer 0x124e5a0002
```

They get the error above. This seems to be related to the fact that these tables are partitioned, because if I try to restore the same dataset without partitioned it successfully imports: 

```
root@:26257/defaultdb> RESTORE database <DB> from 'nodelocal:///anz-cluster-demo-staging-catalogue.sql';
        job_id       |  status   | fraction_completed | rows  | index_entries | system_records |  bytes
+--------------------+-----------+--------------------+-------+---------------+----------------+---------+
  461768851681116161 | succeeded |                  1 | 26001 |          7481 |              0 | 6230861
(1 row)

Time: 201.975ms
```

If instead we try to use `RESTORE` and try table by table we get the following: 

```
root@:26257/defaultdb> RESTORE db.t1 from 'nodelocal:///file.sql';
pq: cannot restore table "t1" without interleave child table 460
root@:26257/defaultdb> RESTORE db.t2 from 'nodelocal:///file.sql';
pq: cannot restore table "t2" without referenced table 459 (or "skip_missing_foreign_keys" option)
root@:26257/defaultdb> RESTORE db.t3 from 'nodelocal:///file.sql' WITH skip_missing_foreign_keys;
pq: cannot restore table "t3" without interleave parent 459
```

I then tried to use IMPORT PGDUMP since we treat our dump files like PGDUMP files and they had provided one as well. 

```
import pgdump 'nodelocal:///file.sql';
pq: interleaved not supported
HINT: This feature is not yet implemented in CockroachDB.
```

Trying to import tables individually using `PGDUMP` we run into the same issue with interleaved not being supported. 

There are some related issues about support for interleaved tables #35462 
And also older issues regarding `could not rewrite span start key` errors #34099, though this seems like it was something different. 

Dump files and backups can be found [here](https://drive.google.com/drive/folders/1YfuLpUuAP_zrwPmuh5MF0ZffDSNjypX6?usp=sharing)

CC @tbg @knz Not sure if you two are able to look into this as it it rather urgent, perhaps you would be able to at least make some headway before more of our team comes in tomorrow morning. Thanks!

**Comment by jordanlewis** [comment]
It looks like there are a few different problems to unpack here.

The first error is produced by the `findDecimalTerminator` method:

```
// findDecimalTerminator finds the decimalTerminator in the given slice.
func findDecimalTerminator(buf []byte) (int, error) {
	if idx := bytes.IndexByte(buf, decimalTerminator); idx != -1 {
		return idx, nil
	}
	return -1, errors.Errorf("did not find terminator %#x in buffer %#x", decimalTerminator, buf)
}
```

I'm not sure whether this is related to the interleaved index problem or not. It might be something distinct to do with a bad decimal encoding.

I believe that the second problem, 

```
import pgdump 'nodelocal:///file.sql';
pq: interleaved not supported
HINT: This feature is not yet implemented in CockroachDB.
```

is less of a big deal. I think it comes down to the fact that we don't support specifying interleaved indexes inline within `CREATE TABLE` statements. I think if we taught `dump` to print separate `CREATE INDEX` statements for interleaved indexes, this problem would go away. And, as a workaround, we could modify the `file.sql` file to specify these `CREATE INDEX` statements separately for the interleaved ones.

I've filed a separate issue for that, #38311, including a workaround.

**Comment by tbg** [comment]
> The first error is produced by the findDecimalTerminator method:

Pretty sure this isn't about decimal, but about bytes encoding. Unfortunately the exact error is produced in three places:

decimal:

https://github.com/cockroachdb/cockroach/blob/7e2ceae0c42fa5b7f32f6042c443358401a77279/pkg/util/encoding/decimal.go#L397

bytes:

https://github.com/cockroachdb/cockroach/blob/7e2ceae0c42fa5b7f32f6042c443358401a77279/pkg/util/encoding/encoding.go#L593

bytesLen:

https://github.com/cockroachdb/cockroach/blob/7e2ceae0c42fa5b7f32f6042c443358401a77279/pkg/util/encoding/encoding.go#L625

prettyPrintInvertedIndexKey:

https://github.com/cockroachdb/cockroach/blob/7e2ceae0c42fa5b7f32f6042c443358401a77279/pkg/util/encoding/encoding.go#L647

The "could not rewrite span key" error wrapping happens in two places, both in `rewriteBackupSpanKey`

https://github.com/cockroachdb/cockroach/blob/a9802cd99366b23b52a2863319b7f03c0c8279dd/pkg/ccl/backupccl/restore.go#L1007-L1020

`RewriteKey` transitively calls `bytesLen`:

https://github.com/cockroachdb/cockroach/blob/e7f6c936a7a0b6c86ef47c7d4a77a3718f557f46/pkg/ccl/storageccl/key_rewriter.go#L174

Now `0x124e5a0002` is really just "NZ" bytes encoded and `.PrefixEnd()`ed, as this test shows:

```go
func TestFoo(t *testing.T) {
	b, err := hex.DecodeString("124e5a0001")
	if err != nil {
		t.Fatal(err)
	}
	t.Errorf("%q", b)
	rdr, dec, err := DecodeBytesAscending(b, nil)
	if err != nil {
		t.Fatal(err)
	}
	t.Errorf("dec=%q rdr=%q", dec, rdr)
}
```

```
--- FAIL: TestFoo (0.00s)
    encoding_test.go:40: "\x12NZ\x00\x01"
    encoding_test.go:45: dec="NZ" rdr=""
```

(I guess `\x12` is the marker, and `\x00\x01` is the terminator for the string)

I'm pretty sure the caller is `PeekLength` because

```go
	b, err := hex.DecodeString("124e5a0002")
	if err != nil {
		t.Fatal(err)
	}
	n, err := PeekLength(b)
	if err != nil {
		t.Fatal(err)
	}
```

exactly reproduces the error, and the other methods emitting that error actually look for that byte in the buffer that they print, which *does* contain it. `PeekLength` skips over the contents and at the end trips over the botched `\x00\x01` escape which has become `\x00\x02` due to `PrefixEnd()`.

At the end of all this, we're feeding garbled keys to the rewriter. I'd be surprised if that isn't just an artifact of how interleaving works and how the rewriter isn't aware (enough) of it.

**Comment by tbg** [comment]
cc @dt

**Comment by dt** [comment]
Thanks for digging into this @tbg!

I dug a bit further and I think I see what is going on.

1. We have spans generated by calling PrefixEnd() on an encoded string:
    a. When we have a table that uses `PARTITION BY LIST`, we create splits for each listed value, using `PrefixEnd` in the process.
    b. `PrefixEnd` mutates the trailing byte(s) of its argument to increment it one.
    c. The partitioning column is a string, and strings are encoded with a magic byte trailer as you pointed  out of \x0001. When the partitioning code makes that into a span, it turns it into \0002.
2. In `RESTORE`'s key rewriter that replaces the old table IDs with new ones wherever they occur in a key, if a table has interleaved children, we decode all the values in the keys as much as needed to determine their length so we can seek past them and look for any table IDs belonging to interleave children.
3. `RESTORE` uses its key rewriter to rewrite the _spans_ in the backup, to get new spans pre-splitting and dividing up the ingestion work.

Thus, when the backed up table was a) partitioned by list on b) a string column and c) had interleaved children, the span rewriting code would try to decode a value in span that had been mutated by PrefixEnd such that it was no longer a valid value encoding and would fail.

I think the fix could actually be as simple as just returning success in the rewriting if value decoding fails, just assuming it is because the suffix has been messed with by span logic but if it this value is the suffix, it also means there are no child table ids to keep looking for. I made it a tiny bit more complicated in https://github.com/cockroachdb/cockroach/pull/38341, but confirmed that with that change, I can indeed restore both backups linked in the issue.

**Comment by tim-o** [comment]
Zendesk ticket #[3432](https://cockroachdb.zendesk.com/agent/tickets/3432) has been linked to this issue.

**Comment by dt** [comment]
38341 fixes the `RESTORE` bug

**Comment by dt** [comment]
Fixed #38341
